File: android/src/main/java/expo/modules/libvlcplayer/managers/MediaPlayerManager.kt
package expo.modules.libvlcplayer.managers

import expo.modules.kotlin.AppContext
import expo.modules.libvlcplayer.LibVlcPlayerView
import java.util.Collections
import java.util.WeakHashMap

object MediaPlayerManager {
    lateinit var audioFocusManager: AudioFocusManager
    lateinit var keepAwakeManager: KeepAwakeManager
    lateinit var pictureInPictureManager: PictureInPictureManager

    val expoViews: MutableSet<LibVlcPlayerView> = Collections.newSetFromMap(WeakHashMap())

    fun registerExpoView(view: LibVlcPlayerView) {
        expoViews.add(view)
    }

    fun unregisterExpoView(view: LibVlcPlayerView) {
        expoViews.remove(view)
    }

    fun onModuleCreate(appContext: AppContext) {
        if (!this::audioFocusManager.isInitialized) {
            audioFocusManager = AudioFocusManager(appContext)
        }

        if (!this::keepAwakeManager.isInitialized) {
            keepAwakeManager = KeepAwakeManager(appContext)
        }

        if (!this::pictureInPictureManager.isInitialized) {
            pictureInPictureManager = PictureInPictureManager(appContext)
        }
    }

    fun onModuleDestroy() {
        expoViews.forEach { view ->
            view.destroyPlayer()
        }
    }

    fun onModuleForeground() {
        expoViews.forEach { view ->
            view.isInBackground = false
            view.onForeground(Unit)
            view.cancelPauseJob()
        }
    }

    fun onModuleBackground() {
        expoViews.forEach { view ->
            view.isInBackground = true
            view.onBackground(Unit)
            view.pauseJob()
        }
    }
}

File: ios/Managers/MediaPlayerManager.swift
import Foundation

class MediaPlayerManager {
    static let shared = MediaPlayerManager()

    let localNetworkManager = LocalNetworkManager()
    let audioSessionManager = AudioSessionManager()
    let keepAwakeManager = KeepAwakeManager()

    let expoViews = NSHashTable<LibVlcPlayerView>.weakObjects()

    func registerExpoView(_ view: LibVlcPlayerView) {
        expoViews.add(view)
    }

    func unregisterExpoView(_ view: LibVlcPlayerView) {
        expoViews.remove(view)
    }

    func onModuleDestroy() {
        for view in expoViews.allObjects {
            view.destroyPlayer()
        }
    }

    func onModuleForeground() {
        for view in expoViews.allObjects {
            view.onForeground()
        }
    }

    func onModuleBackground() {
        for view in expoViews.allObjects {
            view.onBackground()
            view.pauseIf()
        }
    }
}
